### ISSUE TITLE
Enable syntax highlighting for inline code

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#1043]
Thus far, we've always assumed that source code is in a block. However, it's perfectly reasonable to have inline snippets of code that would be nice to have syntax highlighted. 

Perhaps we can recognize the role "source" on monospaced text that hints that it is source and should be syntax highlighted. For example:

```
The assignment [source java]`a += 1` increments the variable `a` by 1.
```

**Comment by nh2** [comment]
> Mardown losers

I guess you can't say that until we actually have this feature.

**Comment by nh2** [comment]
Which, 7 months later, is not yet the case (I'd certainly like to use it).

**Comment by mojavelinux** [comment]
I'll point out that this can be done using an inline extension. You'd probably want to implement it using a dedicated inline macro such as:

```
source:java[a += 1]
```

...though an inline macro can be any regular expression.

**Comment by mojavelinux** [comment]
You could even shorten this to:

```
src:java[a += 1]
```

**Comment by nawroth** [comment]
I think using roles makes the most sense if you roll your own, as the output still looks ok for example when editing in Atom. The inline macro would be great if included in Asciidoctor though.

BTW I have roles like that working in DocGist, but didn't want to push it as it's non-standard. I could push it without telling anyone maybe ;-)

**Comment by mojavelinux** [comment]
I think the idea of roles is that you don't need to worry about them being standard. I'd say it's standard to use them. The syntax encourages it.

**Comment by nawroth** [comment]
What would be a good syntax for something like "in the following paragraph, highlight everything that is monospace as SQL"? Or would that be too much magic?

**Comment by mojavelinux** [comment]
I think that brings us back to the macro, which I still think is a much better choice for this sort of thing because it is much more clear.

```
Type the src:[SELECT] command.
```

Then we can leverage the `source-language` attribute to determine a default language.

Of course, it would still be possible to leverage the `source-language` attribute with:

```
Type the [src]`SELECT` command.
```

(I like src instead of lang as it's more explicit).

If you really want to take it further and automatically highlight monospaced text in a paragraph, I would simply recommend using a postprocessor to detect language keywords in monospaced text and highlight them. This is similar to AsciiDoc Python's specialwords with a bit more context.

**Comment by mojavelinux** [comment]
> as the output still looks ok for example when editing in Atom...

We could easily add this type of thing to Atom. I don't think we should avoid syntax because of what Atom currently supports. We have control over that package.

**Comment by mojavelinux** [comment]
> BTW I have roles like that working in DocGist, but didn't want to push it as it's non-standard. I could push it without telling anyone maybe ;-)

I think it's perfectly fine to experiment with ideas like this in tools. In fact, I encourage it. It helps us get feedback before we spend the time to put it into core. I'd just recommend clearly stating that it is an experimental feature.

**Comment by AlexCzar** [comment]
Is there any implementations of this floating around?

**Comment by nawroth** [comment]
@AlexCzar Adding this to HTML output uing roles is trivial, I just did that for DocGist: http://gist.asciidoctor.org/#_source_code_highlighting

@mojavelinux I like the `source-language` approach. The specialwords approach is too limited and too magic at the same time!

**Comment by mojavelinux** [comment]
> The specialwords approach is too limited and too magic at the same time!

I tend to agree, which is why I actually never implemented that built-in behavior in Asciidoctor. If there's one thing that upsets writers consistently across the board it is when unexpected replacements happen. So keeping the magic to a minimum is essential. Although we like to try to minimize the number of characters to type, I think authors most appreciate syntax that is explicit and clear. In other words, when given the choice between a few extra characters and clarity, I think most writers like clarity. (they just hate XML).

**Comment by madduci** [comment]
is there a way to style also the inline code blocks divided by ``? (e.g., `my-code-here`)

**Comment by mojavelinux** [comment]
Anything is possible with an extension. Right now, no inline source highlighting is provided out of the box.

**Comment by ricpelo** [comment]
What is the easiest approach in order to get this working for HTML and PDF outputs? Sorry, I'm a totally newbie but I'm fascinated with Asciidoctor and its possibilities.

**Comment by mojavelinux** [comment]
You can override the inline_quoted handler (either using a custom converter or a template). Check for monospaced with a particular role and pass that string through the source highlighter.

See https://github.com/asciidoctor/asciidoctor/blob/a34fc3ec665c25cd40ae692265d080dfa03ac851/lib/asciidoctor/converter/html5.rb#L1132-L1145

(Due to the overhead this would create, I think it might be best as a user extension).

**Comment by ricpelo** [comment]
Any news about this?

**Comment by pintergreg** [comment]
I personally like the `source:java[a += 1]` syntax and I am also eager to see this feature in action.

**Comment by ateshuseyin** [comment]
`a +=1`java

**Comment by MattBlissett** [comment]
> You can override the inline_quoted handler (either using a custom converter or a template).

There's a partial implementation of that [https://github.com/gbif/gbif-asciidoctor-toolkit/blob/ed16d2ae21156900139054cc08b67d9616f35b5c/GbifHtmlConverter.rb#L69-L109](here).

1. It's necessary to edit syntax_highlighter.rb to allow it to run on inline nodes, rather than only blocks, see https://github.com/gbif/gbif-asciidoctor-toolkit/blob/8042a71891c391dff2a1e37938e13f33b41f258d/inline-syntax-highlighting.patch

2. (Also in that patch) it's also necessary to edit the CSS base selector, from `pre.pygments` to just `.pygments`, as we can't have a `pre` element inline.

I'm using roles, like this:
```
Here's some [source javascript]`code` or [.source.javascript]`code`
```

----

The next challenge is matching the syntax highlighting in an inline quote. If I have:

```javascript
some.formatted("javascript")
```
and then the document author would like to write `formatted` inline, and have that be the same colour, i.e. purple.

**Comment by mojavelinux** [comment]
> it's also necessary to edit the CSS base selector, from pre.pygments to just .pygments, as we can't have a pre element inline.

I'd be open to this change, though it will require verifying that the rules are still strong enough. Worse case, we may have to duplicate the rules under `code.pygments`. (We might even consider moving the pygments class down to the code element so it's always on the code element).

> It's necessary to edit syntax_highlighter.rb to allow it to run on inline nodes, rather than only blocks,

You don't have to monkeypatch Asciidoctor. You can extend the Asciidoctor::SyntaxHighlighter.for('pygments'), register it for 'pygments', and override the `format` method.

> I'm using roles

Roles looks nice, though that's a whole lot to type. I'm not sure what would be better, though.

> The next challenge is matching the syntax highlighting in an inline quote.

I don't understand this scenario. Could you clarify with a more complete example?

**Comment by MattBlissett** [comment]
> You don't have to monkeypatch Asciidoctor.

Thanks! I've not used Ruby before, I'll tidy it up later.

> > The next challenge is matching the syntax highlighting in an inline quote.
> 
> I don't understand this scenario. Could you clarify with a more complete example?

I think I've found an easy way around this. The author's original document had the first sentence, with `genus` unquoted so it wasn't highlighted as a Javascript string. Replacing it with `"genus"` is hopefully acceptable:

![Screenshot_20190427_122257](https://user-images.githubusercontent.com/583320/56848352-33118b80-68e8-11e9-830f-36e2331b4d24.png)

**Comment by Mogztter** [comment]
> Of course, it would still be possible to leverage the source-language attribute with:
> Type the [src]`SELECT` command.

I quite like this syntax but how do you define the language in the above example?
````
[src=sql]`SELECT`
````
````
[src,lang=sql]`SELECT`
````

**Comment by Mogztter** [comment]
I'm experimenting with the inline macro approach `src:java[a += 1]` but the content between the brackets (attributes) is processed (more specifically specialchars and replacements substitutions)

It can be an issue when the inline code contains specialchars, for instance:

```adoc
src:cypher[MATCH (p:Product)<-[:BOUGHT_BY\]-(:Customer 'ABCCO') RETURN p.productName]
```
Please note that I need to escape the close bracket `\]`.
Also, in the above example, `<-` will be replaced by `&#8592;`.
If I escape the symbol "<-" then `<` will still be replaced by `&lt;`:

```adoc
src:cypher[MATCH (p:Product)\<-[:BOUGHT_BY\]-(:Customer 'ABCCO') RETURN p.productName]
```

As a workaround it's possible to declare an attribute with a passthrough:

```adoc
:cypher-query-bought-by: pass:m[MATCH (p:Product)<-[:BOUGHT_BY\]-(:Customer 'ABCCO') RETURN p.productName]

src:cypher[{cypher-query-bought-by}]
```

So using a monospaced text using a role with a common prefix is probably a better solution (or at least requires less work):

```
[src-cypher]`MATCH (p:Product)<-[:BOUGHT_BY]-(:Customer 'ABCCO') RETURN p.productName`
```

Another benefit is that even if the extension is not loaded, the inline code will use a monospaced font.

**Comment by danyill** [comment]
I was discussing inline syntax highlighting today with Dan and we realised that now the syntax highlighter is pluggable in Asciidoctor 2, much of the work has been done for having an `inline_format` as well as a `format` method.

Thinking out loud, Dan thought 

```
`[.source.ruby]`puts "Hello, World!"` 
```

might be the right kind of thing to avoid a macro.

**Comment by 8thHalfHour** [comment]
This is awesome, you guys! It is December now, so it still on the roadmap?

**Comment by DavidGamba** [comment]
```
`[.source.ruby]`puts "Hello, World!"` 
```
Might be hard for a non regex parser to parse, when you find the leading backtick you capture until you find another one and then you parse the section inside, this 3 backtick syntax throws a wrench on that.

**Comment by mojavelinux** [comment]
@DavidGamba A change to the syntax is not necessary. Inline phrases already support roles. This is just a matter of what the syntax highlighter adapter can handle. In fact, when using highlight.js, it's already possible to do if you customize the adapter. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/syntax_highlighter/highlightjs.rb#L28

**Comment by DavidGamba** [comment]
Thanks, that 3 backtick comment confused me.

**Comment by danyill** [comment]
If you need a solution now, I created an inline macro extension for this in asciidoctor.js which may be of some assistance:

See https://gitlab.com/danyill/asciidoctor-quick-reference/-/blob/122fb996fedf32aebd8798a32d8de5ba311e3515/base/inline-code.js

I used prism.js and incorporated it directly into asciidoctor-web-pdf using a template (see https://gitlab.com/danyill/asciidoctor-quick-reference/-/blob/122fb996fedf32aebd8798a32d8de5ba311e3515/base/template.js#L44) but for general Asciidoctor usage this should be done with a docinfo file or (better) a Docinfo processor

I then use the macro with `src:<lang>[<expression>]:` e.g. `src:asciidoc[\footnote:id[See the identifier\]]`

A fair amount of escaping is required for my purposes because I'm trying to document Asciidoc (and if you have square brackets these need to be escaped).

I'm sure my macro could be improved 😉

**Comment by keryell** [comment]
As explained in https://discuss.asciidoctor.org/Implementing-the-equivalent-of-LaTeX-lstinline-td8444.html#a8446 I am interested by this feature to replace the LaTeX  `\lstinline` macro.
`\lstinline` (described on page 4 of http://mirrors.ctan.org/macros/latex/contrib/listings/listings.pdf) provides 2 different syntax:
- `\lstinline{some code}`;
- and when the code has some `{}` in it: `\lstinline|some code|`, where `|` is actually any character not appearing in `some code`.

The latter syntax is super interesting since you can avoid the escaping nightmare just by picking a character not appearing in the code snippet.
I would suggest we have also such a similar simple but efficient syntax.

**Comment by rmannibucau** [comment]
Any reason to not have a code macro *built-in*, I would reverse it to get the standard options mode: `code:my inline code[.role,....]`?
